src/compiler/checker.ts

```diff
@@ -23074,14 +23103,14 @@ namespace ts {
             return type.flags & TypeFlags.UnionOrIntersection ? every((type as UnionOrIntersectionType).types, f) : f(type);
         }
 
-        function filterType(type: Type, f: (t: Type) => boolean): Type {
-            if (type.flags & TypeFlags.Union) {
-                const types = (type as UnionType).types;
-                const filtered = filter(types, f);
-                if (filtered === types) {
-                    return type;
-                }
-                const origin = (type as UnionType).origin;
+        function filterUnionOrIntersectionType(type: UnionOrIntersectionType, f: (t: Type) => boolean): Type {

```

Comment:
I'd probably make a separate helper for filtering intersections - if you even still need it after swapping to using `getSpreadType` - rather then repurposing `filterType`. most of `filterType`'s callers definitely don't want to filter intersection constituents.

Reply:
Ok, hopefully, the change won't be needed altogether with the introduction of `getSpreadType` (I use it in one other place, but it isn't necessary I think)